src/components/app.tsx

```diff
@@ -89,6 +103,60 @@ class AppBase extends React.PureComponent<Props, {}> {
     });
   }
 
+  private setSpec(spec: Object) {
+    const ajv = new Ajv({
+      validateSchema: true,
+      allErrors: true,
+      extendRefs: 'fail'
+    });
+    ajv.addMetaSchema(draft4Schemas, 'http://json-schema.org/draft-04/schema#');
+
+    const validateVl = ajv.compile(vlSchema);
+    const valid = validateVl(spec);
+
+    if (!valid) {
+      throw new Error("Invalid Schema:" + validateVl.errors.toString());
+    }
+
+    const validSpec: TopLevel<FacetedCompositeUnitSpec> = spec as TopLevel<FacetedCompositeUnitSpec>;
+
+
+    if (!isUnitSpec(validSpec)) {
+      throw new Error("Voyager does not support layered or multi-view vega-lite specs");
+    }
+
+    if (validSpec.data) {
+      const setDataAction = this.setData(validSpec.data) as Promise<any>;
```

Comment:
not sure why do you have to cast `as Promise<any>` here.   

I wonder if we should just void casting by just doing
```
this.setData(validSpec.data)
  .then(...
```